core/src/main/scala/kafka/consumer/BaseConsumer.scala

```diff
@@ -91,6 +94,26 @@ class NewShinyConsumer(topic: Option[String], partitionId: Option[Int], offset:
     }
   }
 
+  def seekToRealPositionsBeforeExit() {
+    while (recordIter.hasNext) {
+      val record = recordIter.next()
+      if (!consumedOffsets.exists(tp => tp._1.topic() == record.topic() && tp._1.partition() == record.partition())) {
```

Comment:
We only want to insert in the map the smallest offset for each partition. Also, maybe a better name for this collection is `smallestUnconsumedOffsets` or something like that.

Reply:
@hachikuji  Can we safely assert that the firstly encountered record for some partition must be of the smallest offset? If yes, we don't need to check the offset things. Instead we push that record's offset to the map. In my test env, it is indeed that the first read record per partition has the smallest offset. Does it make sense?